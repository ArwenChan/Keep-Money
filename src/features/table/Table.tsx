import React, { ChangeEvent, ReactNode } from 'react'
import clsx from 'clsx'
import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableContainer from '@material-ui/core/TableContainer'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import TableSortLabel from '@material-ui/core/TableSortLabel'
import Toolbar from '@material-ui/core/Toolbar'
import Typography from '@material-ui/core/Typography'
import Paper from '@material-ui/core/Paper'
import Checkbox from '@material-ui/core/Checkbox'
import IconButton from '@material-ui/core/IconButton'
import Tooltip from '@material-ui/core/Tooltip'
import DeleteIcon from '@material-ui/icons/Delete'
import { Bill } from '../../utils/dataHandler'
import TextField from '@material-ui/core/TextField'
import classes from './Table.module.scss'

function descendingComparator<T>(a: T, b: T, orderBy: keyof T) {
  if (b[orderBy] < a[orderBy]) {
    return -1
  }
  if (b[orderBy] > a[orderBy]) {
    return 1
  }
  return 0
}

type Order = 'asc' | 'desc'

function getComparator<Key extends keyof any>(
  order: Order,
  orderBy: Key
): (
  a: { [key in Key]?: number | string },
  b: { [key in Key]?: number | string }
) => number {
  return order === 'desc'
    ? (a, b) => descendingComparator(a, b, orderBy)
    : (a, b) => -descendingComparator(a, b, orderBy)
}

function stableSort<T>(array: T[], comparator: (a: T, b: T) => number) {
  const stabilizedThis = array.map((el, index) => [el, index] as [T, number])
  stabilizedThis.sort((a, b) => {
    const order = comparator(a[0], b[0])
    if (order !== 0) return order
    return a[1] - b[1]
  })
  return stabilizedThis.map((el) => el[0])
}

interface HeadCell {
  isNumber: boolean
  id: keyof Bill
  label: string
}

const headCells: HeadCell[] = [
  { id: 'category', isNumber: false, label: '分类' },
  { id: 'amount', isNumber: true, label: '金额' },
  { id: 'time', isNumber: true, label: '时间' },
]

interface EnhancedTableHeadProps {
  numSelected: number
  onRequestSort: (
    event: React.MouseEvent<unknown>,
    property: keyof Bill
  ) => void
  onSelectAllClick: (event: React.ChangeEvent<HTMLInputElement>) => void
  order: Order
  orderBy: string
  rowCount: number
  filter: ReactNode
}

function EnhancedTableHead(props: EnhancedTableHeadProps) {
  const {
    onSelectAllClick,
    order,
    orderBy,
    numSelected,
    rowCount,
    onRequestSort,
    filter: Filter,
  } = props
  const createSortHandler =
    (property: keyof Bill) => (event: React.MouseEvent<unknown>) => {
      onRequestSort(event, property)
    }

  return (
    <TableHead>
      <TableRow>
        <TableCell padding="checkbox">
          <Checkbox
            indeterminate={numSelected > 0 && numSelected < rowCount}
            checked={rowCount > 0 && numSelected === rowCount}
            onChange={onSelectAllClick}
            inputProps={{ 'aria-label': 'select all' }}
          />
        </TableCell>
        {headCells.map((headCell) => (
          <TableCell
            key={headCell.id}
            align={headCell.isNumber ? 'right' : 'left'}
            padding="default"
            sortDirection={orderBy === headCell.id ? order : false}
          >
            {headCell.isNumber ? (
              <TableSortLabel
                active={orderBy === headCell.id}
                direction={orderBy === headCell.id ? order : 'asc'}
                onClick={createSortHandler(headCell.id)}
              >
                {headCell.label}
              </TableSortLabel>
            ) : (
              Filter
            )}
          </TableCell>
        ))}
      </TableRow>
    </TableHead>
  )
}

interface EnhancedTableToolbarProps {
  numSelected: number
  onDelete: () => void
  onChangeMonth: (event: ChangeEvent) => void
}

const EnhancedTableToolbar = (props: EnhancedTableToolbarProps) => {
  const { numSelected, onDelete, onChangeMonth } = props

  return (
    <Toolbar
      className={clsx(classes.toolbarRoot, {
        [classes.toolbarHighlight]: numSelected > 0,
      })}
    >
      {numSelected > 0 ? (
        <Typography
          className={classes.toolbarTitle}
          color="inherit"
          variant="subtitle1"
          component="div"
        >
          {numSelected} selected
        </Typography>
      ) : (
        <Typography
          className={classes.toolbarTitle}
          variant="h6"
          id="tableTitle"
          component="div"
        >
          Keep Money
        </Typography>
      )}
      {numSelected > 0 ? (
        <Tooltip title="Delete">
          <IconButton aria-label="delete" onClick={onDelete}>
            <DeleteIcon />
          </IconButton>
        </Tooltip>
      ) : (
        <TextField
          id="date"
          type="month"
          onChange={onChangeMonth}
          data-testid="calendar"
          inputProps={{ role: 'month' }}
        />
      )}
    </Toolbar>
  )
}

interface EnhancedTableProps {
  data: Array<Bill>
  categoriesMapping: any
  deleteData: (bills: Array<number>) => void
  changeMonth: (event: ChangeEvent) => void
  filter: ReactNode
}

function EnhancedTable(props: EnhancedTableProps) {
  const {
    data: rows,
    categoriesMapping,
    deleteData,
    changeMonth,
    filter,
  } = props
  const [order, setOrder] = React.useState<Order>('asc')
  const [orderBy, setOrderBy] = React.useState<keyof Bill>('time')
  const [selected, setSelected] = React.useState<number[]>([])

  const handleRequestSort = (
    event: React.MouseEvent<unknown>,
    property: keyof Bill
  ) => {
    if (property !== 'category') {
      const isAsc = orderBy === property && order === 'asc'
      setOrder(isAsc ? 'desc' : 'asc')
      setOrderBy(property)
    }
  }

  const handleSelectAllClick = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (event.target.checked) {
      const newSelecteds = rows.map((row) => row.id!)
      setSelected(newSelecteds)
      return
    }
    setSelected([])
  }

  const handleClick = (event: React.MouseEvent<unknown>, key: number) => {
    const selectedIndex = selected.indexOf(key)
    let newSelected: number[] = []
    if (selectedIndex === -1) {
      newSelected = newSelected.concat(selected, key)
    } else if (selectedIndex === 0) {
      newSelected = newSelected.concat(selected.slice(1))
    } else if (selectedIndex === selected.length - 1) {
      newSelected = newSelected.concat(selected.slice(0, -1))
    } else if (selectedIndex > 0) {
      newSelected = newSelected.concat(
        selected.slice(0, selectedIndex),
        selected.slice(selectedIndex + 1)
      )
    }

    setSelected(newSelected)
  }

  const handleDelete = () => {
    deleteData(selected)
    setSelected([])
  }

  const isSelected = (key: number) => selected.includes(key)

  return (
    <div className={classes.root}>
      <Paper className={classes.paper}>
        <EnhancedTableToolbar
          numSelected={selected.length}
          onDelete={handleDelete}
          onChangeMonth={changeMonth}
        />
        <TableContainer>
          <Table
            className={classes.table}
            aria-labelledby="tableTitle"
            size="medium"
            aria-label="bill"
          >
            <EnhancedTableHead
              numSelected={selected.length}
              order={order}
              orderBy={orderBy}
              onSelectAllClick={handleSelectAllClick}
              onRequestSort={handleRequestSort}
              rowCount={rows.length}
              filter={filter}
            />
            <TableBody>
              {stableSort<Bill>(
                rows,
                getComparator<keyof Bill>(order, orderBy)
              ).map((row, index) => {
                const isItemSelected = isSelected(row.id!)
                const labelId = `enhanced-table-checkbox-${index}`

                return (
                  <TableRow
                    hover
                    onClick={(event) => handleClick(event, row.id!)}
                    role="item"
                    aria-checked={isItemSelected}
                    tabIndex={-1}
                    key={row.id}
                    selected={isItemSelected}
                  >
                    <TableCell padding="checkbox">
                      <Checkbox
                        checked={isItemSelected}
                        inputProps={{ 'aria-labelledby': labelId }}
                      />
                    </TableCell>
                    <TableCell
                      component="th"
                      id={labelId}
                      scope="row"
                      padding="none"
                    >
                      {categoriesMapping[row.category]?.name}
                    </TableCell>
                    <TableCell
                      align="right"
                      className={clsx({
                        [classes.outcomeMoney]: row.type === 0,
                      })}
                    >
                      {row.amount}
                    </TableCell>
                    <TableCell align="right">
                      {new Date(Number(row.time)).toLocaleDateString()}
                    </TableCell>
                  </TableRow>
                )
              })}
            </TableBody>
          </Table>
        </TableContainer>
      </Paper>
    </div>
  )
}

export default React.memo(EnhancedTable)
