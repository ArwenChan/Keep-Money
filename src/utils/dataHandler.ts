import Papa from 'papaparse'
import { openDB, readAllRecord, readAllRecordWithKey } from './dbHelper'
const fs = require('fs')
const path = require('path')

export interface Bill {
  id?: number
  type: 0 | 1
  category: string
  amount: number
  time: number
  month: string
}

export interface Category {
  id: string
  name: string
  type: 0 | 1
}

export async function readCSV<T>(filepath: string): Promise<Array<T>> {
  return new Promise((resolve, reject) => {
    if (process.env.NODE_ENV === 'test') {
      const localpath = path.resolve(
        __dirname,
        '../../public/mock/',
        `./${filepath}`
      )
      const file = fs.createReadStream(localpath)
      const csvData: Array<T> = []
      Papa.parse<T>(file, {
        header: true,
        step: function (result) {
          //@ts-ignore, wrong type check here
          csvData.push(result.data)
        },
        complete: function (results) {
          resolve(csvData)
        },
        error: (err) => {
          reject(err)
        },
      })
    } else {
      Papa.parse<T>(filepath, {
        download: true,
        header: true,
        complete: (results) => {
          // console.log(results.data)
          resolve(results.data)
        },
        error: (err) => {
          reject(err)
        },
      })
    }
  })
}

export function getMonth(milliseconds?: number): string {
  const date: Date = milliseconds ? new Date(milliseconds) : new Date()
  return `${date.getFullYear()}${String(date.getMonth() + 1).padStart(2, '0')}`
}

export function getCategoryMapping(categories: Array<Category>) {
  let result: any = {}
  categories.forEach((v) => {
    result[v.id] = { name: v.name, type: v.type }
  })
  return result
}

async function initData(): Promise<[Bill[], Category[]]> {
  return openDB().then(async () => {
    return Promise.all([
      readAllRecordWithKey<Bill>('data'),
      readAllRecord<Category>('category'),
    ])
  })
}

export function computeSummary(data: Bill[]): [number, number] {
  let income = 0
  let outcome = 0
  data.forEach((bill) => {
    if (bill.type === 0) {
      outcome += bill.amount
    } else {
      income += bill.amount
    }
  })
  return [income, outcome]
}

class BaseResource {
  data: any
  status: 'pending' | 'error' | 'success' = 'pending'
  error: any
  promise: Promise<any> | null = null
  read() {
    switch (this.status) {
      case 'pending':
        throw this.promise
      case 'error':
        throw this.error
      default:
        return this.data
    }
  }
}

class AsyncResource extends BaseResource {
  constructor(promise: any) {
    super()
    this.promise = promise
      .then((data: any) => {
        this.status = 'success'
        this.data = data
      })
      .catch((error: any) => {
        this.status = 'error'
        this.error = error
      })
  }
}

export const initResource = new AsyncResource(initData())
