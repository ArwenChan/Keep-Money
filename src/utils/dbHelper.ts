import { Bill, Category, getMonth, readCSV } from './dataHandler'

let db: IDBDatabase
export async function openDB() {
  return new Promise((resolve, reject) => {
    const request = indexedDB.open('keep')
    request.onerror = function (event) {
      console.log('数据库打开报错')
      reject(event)
    }

    request.onsuccess = function (event) {
      db = request.result
      console.log('数据库打开成功')
      resolve(event)
    }

    request.onupgradeneeded = async function (event) {
      console.log('数据库新建')
      db = request.result
      db.createObjectStore('data', { autoIncrement: true }).createIndex(
        'month',
        'month',
        { unique: false }
      )

      db.createObjectStore('category', { keyPath: 'id' })
      const tx = request.transaction
      console.log('downloading...')
      Promise.all([
        readCSV<Bill>('/data/bill.csv'),
        readCSV<Category>('/data/categories.csv'),
      ]).then(([billData, categories]) => {
        tx!.oncomplete = () => {
          initData(billData, categories).then(() => {
            resolve('load data successfully')
          })
        }
      })
    }
  })
}

export async function initData(data: Array<Bill>, categories: Array<Category>) {
  return new Promise((resolve, reject) => {
    const tx = db.transaction(['data', 'category'], 'readwrite')
    const dataScheme = tx.objectStore('data')
    const categoryScheme = tx.objectStore('category')
    data.forEach((value) => {
      value.time = Number(value.time)
      value.month = getMonth(value.time)
      value.type = Number(value.type) as 0 | 1
      value.amount = Number(value.amount)
      dataScheme.add(value)
    })
    categories.forEach((value) => {
      value.type = Number(value.type) as 0 | 1
      categoryScheme.add(value)
    })
    tx.oncomplete = () => {
      resolve('')
    }
    tx.onerror = (event) => {
      reject(event)
    }
  })
}

export async function addData<T>(scheme: string, data: Array<T>) {
  return new Promise((resolve, reject) => {
    const tx = db.transaction(scheme, 'readwrite')
    data.forEach((value) => {
      tx.objectStore(scheme).add(value)
    })
    tx.oncomplete = () => {
      resolve(null)
    }
    tx.onerror = (event) => {
      reject(event)
    }
  })
}

export async function deleteData(scheme: string, ids: string[] | number[]) {
  return new Promise((resolve, reject) => {
    const tx = db.transaction(scheme, 'readwrite')
    ids.forEach((v: string | number) => {
      tx.objectStore(scheme).delete(v)
    })
    tx.oncomplete = () => {
      console.log('删除成功')
      resolve(null)
    }
    tx.onerror = (event) => {
      reject(event)
    }
  })
}

export async function queryIndex<T>(
  scheme: string,
  index: string,
  indexValue: any
): Promise<Array<T>> {
  return new Promise((resolve, reject) => {
    const tx = db.transaction(scheme, 'readonly')
    const request = tx
      .objectStore(scheme)
      .index(index)
      .openCursor(IDBKeyRange.only(indexValue))
    let datas: any = []
    request.onsuccess = function (event) {
      let cursor = request.result
      if (cursor) {
        datas.push({ ...cursor.value, id: cursor.primaryKey })
        cursor.continue()
      } else {
        resolve(datas)
      }
    }
    request.onerror = function (e) {
      reject(e)
    }
  })
}

export async function readAllRecord<T>(table: string): Promise<Array<T>> {
  return new Promise((resolve, reject) => {
    let transaction = db.transaction(table)
    let objectStore = transaction.objectStore(table)
    let request = objectStore.getAll()
    request.onsuccess = function (event) {
      resolve(request.result)
    }

    request.onerror = function (event) {
      reject(event)
    }
  })
}

export async function readAllRecordWithKey<T>(
  table: string
): Promise<Array<T>> {
  return new Promise((resolve, reject) => {
    let transaction = db.transaction(table, 'readonly')
    let objectStore = transaction.objectStore(table)
    let request = objectStore.openCursor()
    let datas: any = []
    request.onsuccess = function (event) {
      let cursor = request.result
      if (cursor) {
        datas.push({ ...cursor.value, id: cursor.primaryKey })
        cursor.continue()
      } else {
        resolve(datas)
      }
    }

    request.onerror = function (event) {
      reject(event)
    }
  })
}
