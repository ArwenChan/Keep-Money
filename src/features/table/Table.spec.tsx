import { fireEvent, render, screen, within } from '@testing-library/react'
import { Bill } from '../../utils/dataHandler'
import Table from './Table'

describe('test table', () => {
  it('test table render', () => {
    const data: Array<Bill> = [
      {
        type: 0,
        time: 1561910400000,
        category: '8s0p77c323',
        amount: 5400,
        month: '201907',
        id: 1,
      },
      {
        type: 0,
        time: 1623063290531,
        category: '0fnhbcle6hg',
        amount: 1500,
        month: '202106',
        id: 2,
      },
      {
        type: 0,
        time: 1623083403918,
        category: '0fnhbcle6hg',
        amount: 1500,
        month: '202106',
        id: 3,
      },
    ]
    const categoriesMapping = {
      '8s0p77c323': {
        name: '房贷',
        type: 0,
      },
      '0fnhbcle6hg': {
        name: '房屋租赁',
        type: 0,
      },
    }
    const deleteData = jest.fn()
    const changeMonth = jest.fn()
    const filter = <span>filter</span>
    const props = { data, categoriesMapping, deleteData, changeMonth, filter }
    const { container } = render(<Table {...props} />)
    expect(container).toMatchSnapshot()

    const calendar = screen.getByTestId('calendar')
    expect(calendar).toBeInTheDocument()
    fireEvent.change(within(calendar).getByRole('month'), {
      target: {
        value: '2010-07',
      },
    })
    expect(changeMonth).toBeCalledTimes(1)
    let items = screen.getAllByRole('item')
    expect(items.length).toEqual(3)
    expect(within(items[0]).queryByText('2019-7-1')).not.toBeNull()
    expect(within(items[1]).queryByText('2021-6-7')).not.toBeNull()
    expect(within(items[2]).queryByText('2021-6-8')).not.toBeNull()
    const tableitems = screen.getAllByRole('rowgroup')
    expect(tableitems.length).toEqual(2)
    const tableHeader = tableitems[0]

    // order by time reverse
    fireEvent.click(within(tableHeader).getByText('时间'))
    items = screen.getAllByRole('item')
    expect(items.length).toEqual(3)
    expect(within(items[0]).queryByText('2021-6-8')).not.toBeNull()
    expect(within(items[1]).queryByText('2021-6-7')).not.toBeNull()
    expect(within(items[2]).queryByText('2019-7-1')).not.toBeNull()

    //order by amount
    fireEvent.click(within(tableHeader).getByText('金额'))
    items = screen.getAllByRole('item')
    expect(items.length).toEqual(3)
    expect(within(items[0]).queryByText('1500')).not.toBeNull()
    expect(within(items[1]).queryByText('1500')).not.toBeNull()
    expect(within(items[2]).queryByText('5400')).not.toBeNull()

    //order by amount reverse
    fireEvent.click(within(tableHeader).getByText('金额'))
    items = screen.getAllByRole('item')
    expect(items.length).toEqual(3)
    expect(within(items[0]).queryByText('5400')).not.toBeNull()
    expect(within(items[1]).queryByText('1500')).not.toBeNull()
    expect(within(items[2]).queryByText('1500')).not.toBeNull()

    //select first one
    fireEvent.click(within(items[0]).getByRole('checkbox'))
    expect(screen.queryByTestId('calendar')).toBeNull()
    expect(screen.queryByText('1 selected')).not.toBeNull()

    //select second one
    fireEvent.click(within(items[1]).getByRole('checkbox'))
    expect(screen.queryByText('2 selected')).not.toBeNull()

    // select last one
    fireEvent.click(within(items[2]).getByRole('checkbox'))
    expect(screen.queryByText('3 selected')).not.toBeNull()

    fireEvent.click(within(items[1]).getByRole('checkbox'))
    expect(screen.queryByText('2 selected')).not.toBeNull()

    fireEvent.click(within(items[2]).getByRole('checkbox'))
    expect(screen.queryByText('1 selected')).not.toBeNull()

    fireEvent.click(within(items[0]).getByRole('checkbox'))
    expect(screen.queryByText('1 selected')).toBeNull()
    expect(screen.queryByTestId('calendar')).not.toBeNull()

    //test all select
    const selectAll = screen.getByLabelText('select all')
    fireEvent.click(selectAll)
    expect(screen.queryByTestId('calendar')).toBeNull()
    expect(screen.queryByText('3 selected')).not.toBeNull()
    fireEvent.click(selectAll)
    expect(screen.queryByTestId('calendar')).not.toBeNull()
    expect(screen.queryByText('3 selected')).toBeNull()

    fireEvent.click(selectAll)
    fireEvent.click(screen.getByLabelText('delete'))
    expect(deleteData).toBeCalledTimes(1)
    expect(screen.queryByLabelText('delete')).toBeNull()
  })
})
