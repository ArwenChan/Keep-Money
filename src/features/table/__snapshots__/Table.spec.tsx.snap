// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test table test table render 1`] = `
<div>
  <div
    class="root"
  >
    <div
      class="MuiPaper-root paper MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular toolbarRoot MuiToolbar-gutters"
      >
        <div
          class="MuiTypography-root toolbarTitle MuiTypography-h6"
          id="tableTitle"
        >
          Keep Money
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root"
          data-testid="calendar"
        >
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input"
              id="date"
              role="month"
              type="month"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="MuiTableContainer-root"
      >
        <table
          aria-label="bill"
          aria-labelledby="tableTitle"
          class="MuiTable-root table"
        >
          <thead
            class="MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-paddingCheckbox"
                scope="col"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      aria-label="select all"
                      class="PrivateSwitchBase-input-4"
                      data-indeterminate="false"
                      type="checkbox"
                      value=""
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </span>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft"
                scope="col"
              >
                <span>
                  filter
                </span>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
                scope="col"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiTableSortLabel-root"
                  role="button"
                  tabindex="0"
                >
                  金额
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                    />
                  </svg>
                </span>
              </th>
              <th
                aria-sort="ascending"
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
                scope="col"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiTableSortLabel-root MuiTableSortLabel-active"
                  role="button"
                  tabindex="0"
                >
                  时间
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                    />
                  </svg>
                </span>
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root"
          >
            <tr
              aria-checked="false"
              class="MuiTableRow-root MuiTableRow-hover"
              role="item"
              tabindex="-1"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      aria-labelledby="enhanced-table-checkbox-0"
                      class="PrivateSwitchBase-input-4"
                      data-indeterminate="false"
                      type="checkbox"
                      value=""
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </span>
              </td>
              <th
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                id="enhanced-table-checkbox-0"
                role="cell"
                scope="row"
              >
                房贷
              </th>
              <td
                class="MuiTableCell-root MuiTableCell-body outcomeMoney MuiTableCell-alignRight"
              >
                5400
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
              >
                2019-7-1
              </td>
            </tr>
            <tr
              aria-checked="false"
              class="MuiTableRow-root MuiTableRow-hover"
              role="item"
              tabindex="-1"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      aria-labelledby="enhanced-table-checkbox-1"
                      class="PrivateSwitchBase-input-4"
                      data-indeterminate="false"
                      type="checkbox"
                      value=""
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </span>
              </td>
              <th
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                id="enhanced-table-checkbox-1"
                role="cell"
                scope="row"
              >
                房屋租赁
              </th>
              <td
                class="MuiTableCell-root MuiTableCell-body outcomeMoney MuiTableCell-alignRight"
              >
                1500
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
              >
                2021-6-7
              </td>
            </tr>
            <tr
              aria-checked="false"
              class="MuiTableRow-root MuiTableRow-hover"
              role="item"
              tabindex="-1"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      aria-labelledby="enhanced-table-checkbox-2"
                      class="PrivateSwitchBase-input-4"
                      data-indeterminate="false"
                      type="checkbox"
                      value=""
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </span>
              </td>
              <th
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                id="enhanced-table-checkbox-2"
                role="cell"
                scope="row"
              >
                房屋租赁
              </th>
              <td
                class="MuiTableCell-root MuiTableCell-body outcomeMoney MuiTableCell-alignRight"
              >
                1500
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
              >
                2021-6-8
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
